Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/BaseToolWithRectangleShape.kt

```diff
@@ -0,0 +1,762 @@
+        if (options.all { it != null }) {
+            block(options.filterNotNull())
+        }
+    }
+
+    fun setBitmap(bitmap: Bitmap?) {
+        if (bitmap != null) {
+            drawingBitmap = bitmap
+        }
+        workspace.invalidate()
+    }
+
+    override fun handleDown(coordinate: PointF?): Boolean {
+        movedDistance.set(0f, 0f)
+        coordinate?.apply {
+            previousEventCoordinate = PointF(x, y)
+            currentAction = getAction(x, y)
+        }
+        touchDownPositionX = toolPosition.x
+        touchDownPositionY = toolPosition.y
+        return true
+    }
+
+    override fun handleMove(coordinate: PointF?): Boolean {
+        if (previousEventCoordinate == null || currentAction == null) {
+            return false
+        }
+        ifNotNull(coordinate, previousEventCoordinate) { (coordinate, previousEventCoordinate) ->
+            val delta = PointF(
+                coordinate.x - previousEventCoordinate.x,
+                coordinate.y - previousEventCoordinate.y
+            )
+            movedDistance.set(movedDistance.x + abs(delta.x), movedDistance.y + abs(delta.y))
+            previousEventCoordinate.set(coordinate.x, coordinate.y)
+            when (currentAction) {
+                FloatingBoxAction.MOVE -> move(delta.x, delta.y)
+                FloatingBoxAction.RESIZE -> resize(delta.x, delta.y)
+                FloatingBoxAction.ROTATE -> rotate(delta.x, delta.y)
+                else -> Unit
+            }
+        }
+        return true
+    }
+
+    override fun handleUp(coordinate: PointF?): Boolean {
+        if (previousEventCoordinate == null) {
+            return false
+        }
+        ifNotNull(coordinate, previousEventCoordinate) { (coordinate, previousEventCoordinate) ->
+            movedDistance.x += abs(coordinate.x - previousEventCoordinate.x)
+            movedDistance.y += abs(coordinate.y - previousEventCoordinate.y)
+        }
+        if (CLICK_IN_BOX_MOVE_TOLERANCE >= movedDistance.x && CLICK_IN_BOX_MOVE_TOLERANCE >= movedDistance.y) {
+            toolPosition.x = touchDownPositionX
+            toolPosition.y = touchDownPositionY
+        }
+        return true
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
+    fun boxContainsPoint(coordinate: PointF): Boolean {
+        val relativeToOriginX = coordinate.x - toolPosition.x
+        val relativeToOriginY = coordinate.y - toolPosition.y
+        val radians = -(boxRotation * PI.toFloat() / 180f)
+        val rotatedX =
+            relativeToOriginX * cos(radians) - relativeToOriginY * sin(radians) + toolPosition.x
+        val rotatedY =
+            relativeToOriginX * sin(radians) + relativeToOriginY * cos(radians) + toolPosition.y
+        return rotatedX > toolPosition.x - boxWidth / 2 && rotatedX < toolPosition.x + boxWidth / 2 && rotatedY > toolPosition.y - boxHeight / 2 && rotatedY < toolPosition.y + boxHeight / 2
```

Comment:
I think that's not actually duplicate :)

Reply:
Since you are going to refactor this class again it's not necessary.
Many lines would look much nicer if the box data was within a class Box (or similar), and some operations could be applied to it.

Reply:
hmm.. nice idea, I'll try and do something similar when I refactor it again later.